lib/ood_core/job/adapters/linux_host/launcher.rb

```diff
@@ -0,0 +1,262 @@
+require 'pathname'
+require 'securerandom'
+require 'shellwords'
+require 'time'
+
+# Object used for simplified communication SSH hosts
+#
+# @api private
+class OodCore::Job::Adapters::LinuxHost::Launcher
+  attr_reader :contain, :debug, :site_timeout, :session_name_label, :singularity_bin,
+    :site_singularity_bindpath, :default_singularity_image, :ssh_hosts,
+    :strict_host_checking, :submit_host, :tmux_bin, :username
+  # The root exception class that all LinuxHost adapter-specific exceptions inherit
+  # from
+  class Error < StandardError; end
+
+  UNIT_SEPARATOR = "\x1F"
+
+  # @param debug Whether the adapter should be used in debug mode
+  # @param site_timeout [#to_i] A period after which the job should be killed or nil
+  # @param singularity_bin  Path to the Singularity executable
+  # @param singularity_bindpath A comma delimited string of host paths to bindmount into the guest; sets SINGULARITY_BINDPATH environment variable
+  # @param singularity_image [#to_s] Path to the Singularity image
+  # @param ssh_hosts List of hosts to check when scanning for running jobs
+  # @param strict_host_checking Allow SSH to perform strict host checking
+  # @param submit_host The SSH-able host
+  # @param tmux_bin [#to_s] Path to the tmux executable
+  def initialize(
+    contain: false,
+    debug: false,
+    site_timeout: nil,
+    singularity_bin:,
+    singularity_bindpath: '/etc,/media,/mnt,/opt,/run,/srv,/usr,/var,/users',
+    singularity_image:,
+    ssh_hosts:,
+    strict_host_checking: false,
+    submit_host:,
+    tmux_bin:,
+    **_
+  )
+    @contain = !! contain
+    @debug = !! debug
+    @site_timeout = site_timeout.to_i
+    @session_name_label = 'launched-by-ondemand'
+    @singularity_bin = Pathname.new(singularity_bin)
+    @site_singularity_bindpath = singularity_bindpath.to_s
+    @default_singularity_image = Pathname.new(singularity_image)
+    @ssh_hosts = ssh_hosts
+    @strict_host_checking = strict_host_checking
+    @submit_host = submit_host
+    @tmux_bin = tmux_bin
+    @username = Etc.getlogin
+  end
+
+  # @param hostname [#to_s] The hostname to submit the work to
+  # @param script [OodCore::Job::Script] The script object defining the work
+  def start_remote_session(script)
+    cmd = ssh_cmd(submit_host)
+
+    session_name = unique_session_name
+    output = call(*cmd, stdin: wrapped_script(script, session_name))
+    hostname = output.strip
+
+    "#{session_name}@#{hostname}"
+  end
+
+  def stop_remote_session(session_name, hostname)
+    cmd = ssh_cmd(hostname)
+
+    # discover tmux sessions
+    # discover the PID for the given session
+    # get all children of that PID
+    # find the sinit process for the tmux PID
+    # kill the sinit process
+    kill_cmd = <<~SCRIPT
+    kill $(pstree -p $(tmux list-panes -aF '#\{session_name} \#{pane_pid}' | grep '#{session_name}' | cut -f 2 -d ' ') | grep -Po 'sinit\\(\\d+' | grep -Po '[0-9]+')
+    SCRIPT
+
+    call(*cmd, stdin: kill_cmd)
```

Comment:
As discussed in-person: this change from using `tmux kill-session` is necessary because user scripts with non-trivial activities (like desktops) were not dying when their parent tmux session died; killing the running container works as expected. Simpler ways of doing this are welcome.

Note also that this creates explicit system dependencies on:

- GNU `grep` (I'm going to rewrite one last time to remove the GNU / `-P` requirement)
- `pstree`
- `cut`

Reply:
... which brings the total system dependency list necessary to use this feature to:

- `cut`
- `grep`
- `pstree`
- `Singularity`
- `ssh`
- `tmux`

Reply:
@treydock will you please review one more time? In particular the method to kill sessions changed significantly: https://github.com/OSC/ood_core/pull/151/files#diff-ef23b874dfc17f3318ec7a3985205bbcR69-R89

Reply:
Looks fine to me.

Reply:
Thanks.